packages/browser-repl/src/components/shell.tsx

```diff
@@ -253,235 +287,185 @@ export class _Shell extends Component<ShellProps, ShellState> {
+
+      // don't evaluate empty input, but do add it to the output
+      if (!code || code.trim() === '') {
+        newOutput.push({
+          format: 'input',
+          value: ' ',
+        });
+        capLengthEnd(newOutput, maxOutputLength);
+        onOutputChanged?.(newOutput);
+        return;
+      }
 
-    output = this.addEntriesToOutput([outputLine]);
-    const history = this.addEntryToHistory(code);
-    this.setState({
-      operationInProgress: false,
+      // add input to output
+      newOutput.push({
+        format: 'input',
+        value: code,
+      });
+      capLengthEnd(newOutput, maxOutputLength);
+      onOutputChanged?.(newOutput);
+
+      const outputLine = await evaluate(code);
+
+      // add output to output
+      newOutput.push(outputLine);
+      capLengthEnd(newOutput, maxOutputLength);
+      onOutputChanged?.(newOutput);
+
+      // update history
+      newHistory.unshift(code);
+      changeHistory(
+        newHistory,
+        redactInfo ? 'redact-sensitive-data' : 'keep-sensitive-data'
+      );
+      capLengthStart(newHistory, maxHistoryLength);
+      onHistoryChanged?.(newHistory);
+    },
+    [
       output,
       history,
-    });
-    this.props.onOutputChanged(output);
-    this.props.onHistoryChanged(history);
-  };
-
-  private appendEmptyInput(): void {
-    const inputLine: ShellOutputEntry = {
-      format: 'input',
-      value: ' ',
-    };
-
-    const output = this.addEntriesToOutput([inputLine]);
-
-    this.setState({ output });
-  }
+      onOutputChanged,
+      evaluate,
+      redactInfo,
+      maxHistoryLength,
+      onHistoryChanged,
+      maxOutputLength,
+    ]
+  );
 
-  private scrollToBottom(): void {
-    if (!this.shellInputElement) {
+  const scrollToBottom = useCallback(() => {
+    if (!shellInputElement) {
       return;
     }
 
-    this.shellInputElement.scrollIntoView();
-  }
+    shellInputElement.scrollIntoView();
+  }, [shellInputElement]);
 
-  private onShellClicked = (event: React.MouseEvent): void => {
-    // Focus on input when clicking the shell background (not clicking output).
-    if (event.currentTarget === event.target) {
-      this.focusEditor();
-    }
-  };
-
-  private setEditor = (editor: any | null) => {
-    this.editor = editor;
-  };
-
-  focusEditor = (): void => {
-    this.editor?.focus();
-  };
-
-  private onSigInt = (): Promise<boolean> => {
-    if (
-      this.state.operationInProgress &&
-      (this.props.runtime as WorkerRuntime).interrupt
-    ) {
-      return (this.props.runtime as WorkerRuntime).interrupt();
+  useEffect(() => {
+    scrollToBottom();
+
+    void updateShellPrompt().then(async () => {
+      if (initialEvaluate) {
+        const evalLines = normalizeInitialEvaluate(initialEvaluate);
+        for (const input of evalLines) {
+          await onInput(input);
+        }
+      }
+    });
+  }, [initialEvaluate, onInput, scrollToBottom, updateShellPrompt, output]);
```

Comment:
As you're switching to an effect and as `output` is controlled, this effect will be running every time the output is changed (so every user evaluation). To get a "true" on mount effect, this should have an empty dependency list, usually a way to do this is to ref everything that's being used here

Reply:
`output` shouldn't even used here here. Don't know what I was doing there. I put initialEvaluate in a ref and use that. The others are all callbacks that shouldn't be changing I think - manual testing shows that it isn't. Is this good enough?

Reply:
ok so there were more props being depended on indirectly.. 😢